lab.R

```diff
@@ -32,88 +32,116 @@ data_folder = file.path('data')
 #' 
-
+int_dates2 = int_dates1 %>% 
+    ydm()
 
 
 #' # Problem 3 #
 #' *The dates formatted like "08/15/2018" are easy to parse with the `mdy()` function from lubridate.  Confirm that this works correctly for the instances in `dates`.  The `quiet` argument can be used to suppress the warnings about parsing failures (which we expect here).  Assign the result to `mdy_dates`*
 #' 
+mdy_dates = mdy(dates, quiet = TRUE)
 
 
 
 #' # Problem 4 #
 #' 1. *Now we need to put these two approaches together.  We can do this using `ifelse()` (a base R function), `if_else()` or `case_when()` (`dplyr` functions).  All three provide a "vectorised if": if the value from `mdy_dates` is NA, then use the value from `int_dates2`; otherwise use the value from `mdy_dates`. Assign the result to `parsed_dates`.*
 #' 
-
+parsed_dates = if_else(is.na(mdy_dates),int_dates2, mdy_dates)
+parsed_dates
 #' 2. *We can encapsulate the logic for parsing the dates into a function to create more readable code.  Uncomment the code lines below (leave the lines starting with #') and fill in the blanks with the code you developed above.*
 #' 
 
 #' Parsing dates from the mangled NSF data
 #' @param dates A character vector of dates to be parsed
 #' @return A vector of class `Date`
-# parse_dates = function(dates) {
-#     int_dates = ----
-#         
-#     mdy_dates = ----
-#     
-#     parsed_dates = ----
-#     
-#     return(parsed_dates)
-# }
+parse_dates = function(d) {
+    int_dates = d %>% 
+        as.integer() %>% 
+        as.Date(origin = '1899-12-30') %>% 
+        ydm(quiet = TRUE)
+
+    mdy_dates = mdy(d, quiet = TRUE)
+
+    parsed_dates = if_else(is.na(mdy_dates),int_dates, mdy_dates)
+
+    return(parsed_dates)
+}
+
+parse_dates(dates)
 
 
 
 #' # Problem 5 #
 #' 1. *Now that we have our function to parse/clean the dates, we want to apply it to each of the three input files.  In most languages we would use a for loop to do this.  Because R emphasizes vectors rather than length-one variables, in R the preferred idiom is to use `apply()` or the somewhat more user-friendly variations of `purrr::map()`.  Both of these functions apply a function to every element of a list.  So we'll want to write a single function that takes, as input, a path to a XLSX file; reads it using `read_excel()`; applies `parse_dates()` to the `StartDate` colum; and then returns the cleaned dataframe.  Call this function `read_and_parse()`.* 
 #' 
+read_and_parse = function(filePath) {
+    Excelreader = read_excel(filePath)
+    ParsedDates = data.frame(purrr::map(Excelreader$StartDate, parse_dates))
+        return(ParsedDates)
+}
```

Comment:
We want to use `map()` to run `read_and_parse` on each of the input files, not the elements of `StartDate`.  `parse_dates()` is already vectorized.